From e5150a06351570f050ede4c12706be460780df24 Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Sun, 13 Mar 2022 04:45:08 +0900
Subject: [PATCH 12/12] udev: try to reload selinux label database less
 frequently

Previously, `event_run()` was called repeatedly in one `event_queue_start()`
invocation. Hence, the SELinux label database is reloaded many times needlessly.
Other settings, e.g. udev rules or hwdata, are tried to be reloaded in the
beginning of `event_queue_start()`. Let's also do so for the SELinux database.
---
 src/udev/udevd.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/udev/udevd.c b/src/udev/udevd.c
index 6707befecf..fd93a1a4c1 100644
--- a/src/udev/udevd.c
+++ b/src/udev/udevd.c
@@ -795,10 +795,6 @@ static int event_run(Event *event) {
         /* Re-enable the debug message for the next batch of events */
         log_children_max_reached = true;
 
-        /* fork with up-to-date SELinux label database, so the child inherits the up-to-date db
-         * and, until the next SELinux policy changes, we safe further reloads in future children */
-        mac_selinux_maybe_reload();
-
         /* start new worker and pass initial device */
         r = worker_spawn(manager, event);
         if (r < 0)
@@ -988,6 +984,10 @@ static int event_queue_start(Manager *manager) {
                         return log_warning_errno(r, "Failed to read udev rules: %m");
         }
 
+        /* fork with up-to-date SELinux label database, so the child inherits the up-to-date db
+         * and, until the next SELinux policy changes, we safe further reloads in future children */
+        mac_selinux_maybe_reload();
+
         LIST_FOREACH_SAFE(event, event, event_next, manager->events) {
                 if (event->state != EVENT_QUEUED)
                         continue;
-- 
2.25.1

